core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -58,22 +59,37 @@ class SnappySessionState(snappySession: SnappySession)
   override lazy val sqlParser: SnappySqlParser =
     contextFunctions.newSQLParser(this.snappySession)
 
-  override lazy val analyzer: Analyzer = {
-    new Analyzer(catalog, conf) {
-      override val extendedResolutionRules =
-        new PreprocessTableInsertOrPut(conf) ::
-        new FindDataSourceTable(snappySession) ::
-        DataSourceAnalysis(conf) :: ResolveRelationsExtended ::
-        (if (conf.runSQLonFile) new ResolveDataSource(snappySession) :: Nil else Nil)
-
-      override val extendedCheckRules = Seq(
-        datasources.PreWriteCheck(conf, catalog), PrePutCheck)
-    }
+  override lazy val analyzer: Analyzer = new Analyzer(catalog, conf) {
+    override val extendedResolutionRules =
+      new PreprocessTableInsertOrPut(conf) ::
+          new FindDataSourceTable(snappySession) ::
+          DataSourceAnalysis(conf) ::
+          ResolveRelationsExtended ::
+          ResolveQueryHints(snappySession) ::
+          (if (conf.runSQLonFile) new ResolveDataSource(snappySession) :: Nil else Nil)
+
+    override val extendedCheckRules = Seq(
+      datasources.PreWriteCheck(conf, catalog), PrePutCheck)
   }
 
-  override lazy val optimizer: SparkOptimizer = {
-    new SparkOptimizer(catalog, conf, experimentalMethods) {
-      override def batches: Seq[Batch] = super.batches :+
+  override lazy val optimizer: Optimizer = new SparkOptimizer(catalog, conf, experimentalMethods) {
+    override def batches: Seq[Batch] = {
+      implicit val ss = snappySession
+      var insertedSnappyOpts = 0
+      val modified = super.batches.map {
+        case batch if batch.name.equalsIgnoreCase("Operator Optimizations") =>
+          insertedSnappyOpts += 1
+          val (left, right) = batch.rules.splitAt(batch.rules.indexOf(ReorderJoin))
+          Batch(batch.name, batch.strategy, left ++ Some(ResolveIndex()) ++ right
+              : _*)
```

Comment:
@soubhik-c Looks like ResolveIndex rule is now applied before ReorderJoin? to avoid filter pushdown before ResolveIndex?